database/testing/trackeddb.go

```diff
@@ -17,17 +18,29 @@ var defaultTransactionRunner = txn.NewTransactionRunner()
 
 // trackedDB is used for testing purposes.
 type trackedDB struct {
-	db *sql.DB
+	db *sqlair.DB
 }
 
-func (t *trackedDB) Txn(ctx context.Context, fn func(context.Context, *sql.Tx) error) error {
+// Txn executes the input function against the tracked database, using
+// the sqlair package. The sqlair package provides a mapping library for
+// SQL queries and statements.
+// Retry semantics are applied automatically based on transient failures.
+// This is the function that almost all downstream database consumers
+// should use.
+func (t *trackedDB) Txn(ctx context.Context, fn func(context.Context, *sqlair.TX) error) error {
+	return errors.Trace(defaultTransactionRunner.Txn(ctx, t.db, fn))
```

Comment:
This reminds me. We have a functional options pattern for `NewTransactionRunner`, but the options are never used. 

`trackedDB` only ever uses the singleton `defaultTransactionRunner`, so it's not like we can ever use a different one anyway.

We should probably pass the runner to `trackedDB`. We could use a single instance in a flyweight pattern, which would have the same memory efficiency as the singleton.

Reply:
I actually want to start using a different txn runner based on the namespace in the future. As a new PR, I'm about to put up limits on the concurrency of a given transaction. So we conceivably have a different set of options per namespace.

Reply:
See: https://github.com/juju/juju/pull/15666